src/OpenTelemetry.Api/Context/Propagation/TraceContextPropagator.cs

```diff
@@ -252,14 +313,130 @@ internal static bool TryExtractTracestate(string[] tracestateCollection, out str
 
         private static byte HexCharToByte(char c)
         {
-            if (((c >= '0') && (c <= '9'))
-                || ((c >= 'a') && (c <= 'f'))
-                || ((c >= 'A') && (c <= 'F')))
+            if ((c >= '0') && (c <= '9'))
```

Comment:
Here's the code I use for hex-to-byte:
```csharp
		private static int GetByteValue(char c)
		{
			int val = c - (c < 58 ? 48 : (c < 97 ? 55 : 87));
			return val > 15 || val < 0
				? throw new ArgumentOutOfRangeException(nameof(c), c, $"Invalid character: {c}.")
				: val;
		}
```
I didn't perf test vs what is here but it might be faster. Seems to have less testing and only does subtraction :shrug:

Reply:
The statement would fail for the 10 chars before 'A' and 10 chars before 'a'. For example, if the char is `(char)('a'-1)`, `GetByteValue` would return 9.

Also, I just realized traceparent string only allows lower case, so the upper case can also default to argument exception.

I'll do some perf test later today.

Reply:
I ran a test for multiple code snippets. The `GetByteValue` is only faster for special cases (`1` and `a`). But I prefer the readability and correctness of the `HexCharToByte` method.

|                   Method | c |      Mean |     Error |    StdDev |    Median | Ratio | RatioSD |
|------------------------- |-- |----------:|----------:|----------:|----------:|------:|--------:|
|            HexCharToByte | 0 | 1.0132 ns | 0.0798 ns | 0.1170 ns | 0.9766 ns |  1.00 |    0.00 |
| HexCharToByteMagicNumber | 0 | 2.2548 ns | 0.1007 ns | 0.0942 ns | 2.2535 ns |  2.25 |    0.25 |
|             GetByteValue | 0 | 1.8064 ns | 0.0974 ns | 0.1396 ns | 1.7982 ns |  1.80 |    0.23 |
|       GetByteValueInByte | 0 | 1.7260 ns | 0.0941 ns | 0.0881 ns | 1.7181 ns |  1.73 |    0.22 |
|                          |   |           |           |           |           |       |         |
|            HexCharToByte | 1 | 2.1585 ns | 0.1034 ns | 0.1449 ns | 2.1418 ns |  1.00 |    0.00 |
| HexCharToByteMagicNumber | 1 | 0.9946 ns | 0.0836 ns | 0.1590 ns | 0.9778 ns |  0.47 |    0.08 |
|             GetByteValue | 1 | 1.5450 ns | 0.0912 ns | 0.1447 ns | 1.5105 ns |  0.73 |    0.10 |
|       GetByteValueInByte | 1 | 1.2502 ns | 0.0883 ns | 0.2047 ns | 1.2247 ns |  0.59 |    0.12 |
|                          |   |           |           |           |           |       |         |
|            HexCharToByte | 5 | 1.0440 ns | 0.0796 ns | 0.1797 ns | 1.0173 ns |  1.00 |    0.00 |
| HexCharToByteMagicNumber | 5 | 1.1133 ns | 0.0857 ns | 0.2417 ns | 1.0255 ns |  1.10 |    0.29 |
|             GetByteValue | 5 | 1.6412 ns | 0.0920 ns | 0.0860 ns | 1.6567 ns |  1.61 |    0.29 |
|       GetByteValueInByte | 5 | 1.7342 ns | 0.0805 ns | 0.0861 ns | 1.7294 ns |  1.67 |    0.32 |
|                          |   |           |           |           |           |       |         |
|            HexCharToByte | 9 | 0.9357 ns | 0.0819 ns | 0.0944 ns | 0.9138 ns |  1.00 |    0.00 |
| HexCharToByteMagicNumber | 9 | 2.1795 ns | 0.0889 ns | 0.0951 ns | 2.1800 ns |  2.34 |    0.29 |
|             GetByteValue | 9 | 1.6841 ns | 0.0630 ns | 0.0558 ns | 1.6811 ns |  1.80 |    0.20 |
|       GetByteValueInByte | 9 | 1.2009 ns | 0.0839 ns | 0.1149 ns | 1.2089 ns |  1.29 |    0.15 |
|                          |   |           |           |           |           |       |         |
|            HexCharToByte | a | 2.7667 ns | 0.1139 ns | 0.1356 ns | 2.7555 ns |  1.00 |    0.00 |
| HexCharToByteMagicNumber | a | 1.7536 ns | 0.1004 ns | 0.1885 ns | 1.7337 ns |  0.64 |    0.08 |
|             GetByteValue | a | 2.3222 ns | 0.1427 ns | 0.4118 ns | 2.2874 ns |  0.88 |    0.16 |
|       GetByteValueInByte | a | 2.1082 ns | 0.1062 ns | 0.0887 ns | 2.1193 ns |  0.76 |    0.05 |
|                          |   |           |           |           |           |       |         |
|            HexCharToByte | c | 1.5526 ns | 0.0948 ns | 0.2307 ns | 1.5201 ns |  1.00 |    0.00 |
| HexCharToByteMagicNumber | c | 1.5948 ns | 0.0934 ns | 0.1182 ns | 1.5870 ns |  1.00 |    0.19 |
|             GetByteValue | c | 2.1821 ns | 0.1027 ns | 0.1372 ns | 2.1585 ns |  1.38 |    0.23 |
|       GetByteValueInByte | c | 2.1224 ns | 0.1048 ns | 0.1287 ns | 2.1166 ns |  1.33 |    0.20 |
|                          |   |           |           |           |           |       |         |
|            HexCharToByte | f | 1.6603 ns | 0.0940 ns | 0.1463 ns | 1.6193 ns |  1.00 |    0.00 |
| HexCharToByteMagicNumber | f | 2.7668 ns | 0.1195 ns | 0.2468 ns | 2.7011 ns |  1.71 |    0.23 |
|             GetByteValue | f | 2.0917 ns | 0.0898 ns | 0.1398 ns | 2.0756 ns |  1.27 |    0.15 |
|       GetByteValueInByte | f | 2.1754 ns | 0.1056 ns | 0.1037 ns | 2.1611 ns |  1.35 |    0.13 |


Using [BenchmarkDotNet.Templates](https://www.nuget.org/packages/BenchmarkDotNet.Templates/) for performance measurement, run the code below with `dotnet run -c Release`:

```
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Runtime.CompilerServices;
using BenchmarkDotNet;
using BenchmarkDotNet.Attributes;

namespace gh_1668
{
    public class Benchmarks
    {
        // public field
        [ParamsSource(nameof(InputValues))]
        public char c;

        public static IEnumerable<char> InputValues() => new[]
        {
            '0',
            '1',
            '5',
            '9',
            'a',
            'c',
            'f',
            'z'
        };

        [Benchmark(Baseline = true)]
        public byte HexCharToByte()
        {
            if ((c >= '0') && (c <= '9'))
            {
                return (byte)(c - '0');
            }

            if ((c >= 'a') && (c <= 'f'))
            {
                return (byte)(c - 'a' + 10);
            }

            throw new ArgumentOutOfRangeException(nameof(c), c, $"Invalid character: {c}.");
        }

        [Benchmark]
        public byte HexCharToByteMagicNumber()
        {
            if ((c >= '0') && (c <= '9'))
            {
                return (byte)(c - '0');
            }

            if ((c >= 'a') && (c <= 'f'))
            {
                return (byte)(c - 87);
            }

            throw new ArgumentOutOfRangeException(nameof(c), c, $"Invalid character: {c}.");
        }

        [Benchmark]
        public int GetByteValue()
		{
			int val = c - (c < 58 ? 48 : (c < 97 ? 55 : 87));
			return val > 15 || val < 0
				? throw new ArgumentOutOfRangeException(nameof(c), c, $"Invalid character: {c}.")
				: val;
		}

        [Benchmark]
        public byte GetByteValueInByte()
		{
			byte val = (byte)(c - (c < 58 ? 48 : (c < 97 ? 55 : 87)));
			return val > 15 || val < 0
				? throw new ArgumentOutOfRangeException(nameof(c), c, $"Invalid character: {c}.")
				: val;
		}
    }
}
```